packages/tupaia-web-server/src/routes/EntityAncestorsRoute.ts

```diff
@@ -20,29 +20,36 @@ const DEFAULT_FIELDS = ['parent_code', 'code', 'name', 'type'];
 
 export class EntityAncestorsRoute extends Route<EntityAncestorsRequest> {
   public async buildResponse() {
-    const { params, query, ctx, models } = this.req;
+    const { params, query, ctx, models, accessPolicy } = this.req;
     const { rootEntityCode, projectCode } = params;
     const { includeRootEntity = false, ...restOfQuery } = query;
 
-    const project = (
-      await ctx.services.central.fetchResources('projects', {
-        filter: { code: projectCode },
-        columns: ['config'],
-      })
-    )[0];
-    const { config } = project;
+    const frontendExcludedFilter = await generateFrontendExcludedFilter(
+      models,
+      accessPolicy,
+      projectCode,
+    );
+
+    let shouldIncludeRootEntity = includeRootEntity;
 
-    const { typesExcludedFromWebFrontend } = models.entity;
+    // make sure we don't include the root entity if it's a type that should be excluded
+    if (includeRootEntity) {
+      const excludedTypes = await getTypesToExclude(models, accessPolicy, projectCode);
+      const rootEntity = await ctx.services.entity.getEntity(rootEntityCode, projectCode);
+      if (excludedTypes.includes(rootEntity.type)) {
+        shouldIncludeRootEntity = false;
+      }
+    }
```

Comment:
I wasn't sure where to add this logic tbh - I considered the entity server but it's very Tupaia specific, so I figured this was as good of a place as any. The easiest way was to just not return the root entity if it was an excluded type, because the the moment it always does return.